smart-contract-verifier/src/compiler/fetcher.rs

```diff
@@ -27,11 +28,19 @@ pub enum FetchError {
     File(#[from] std::io::Error),
     #[error("tokio sheduling error: {0}")]
     Schedule(#[from] tokio::task::JoinError),
+    #[error("validation failed: {0}")]
+    Validation(anyhow::Error),
+}
+
+#[async_trait]
+pub trait FileValidator: Send + Sync {
+    async fn validate(&self, ver: &Version, path: &Path) -> Result<(), anyhow::Error>;
 }
 
 #[async_trait]
 pub trait Fetcher: Send + Sync {
     async fn fetch(&self, ver: &Version) -> Result<PathBuf, FetchError>;
+    fn with_validator(&mut self, validator: Arc<dyn FileValidator>);
```

Comment:
Why not just to add a validator as a constructor argument into `ListFetcher::new()` and `S3Fetcher::new()`? It seems that the code would be simpler in that case, while you would get the same result. Additionally, you may accept not `Option<dyn FileValidator>` but `dyn FileValidator` itself into the constructor in that case, which I'm not sure that is better, but still give you the same result with less amount of code and abstractions

Reply:
because I thought that validators should be a part of fetcher interface. Like, you can fetch and while fetching run some custom validator
And I think option is actually still needed, because we might use this fetchers for vyper compilation, for example, and there could be no validation
If we make validator as part of constructor, then if someone will add new fetcher, he could not include validator in it's class, but I think we'd like to require all fetcher to support validators

Reply:
> because I thought that validators should be a part of fetcher interface. Like, you can fetch and while fetching run some custom validator

Yes, but it does not bring to the `Fetcher` trait any additional meaning. You can easily do nothing on that `with_validator` method, as well as to pass `None` to `write_executable` even if some validator has been set with `with_validator`

> And I think option is actually still needed, because we might use this fetchers for vyper compilation, for example, and there could be no validation

Maybe we may just create a `NoOpValidator` which can be used as a default validator for fetching data that could not be validated? I mean, we could leave the `Option` as well, but removing it may make the code more clear. Additionally, `write_executable` would require to pass `FileValidator` as an argument, so that all fetchers that uses that method (supposedly, all **should**) would be forced to pass something as a validator here and, thus, think about what it should be.

Reply:
I think `Option<dyn Validator>` is much clearer that `dyn Validator = NoOpValidator`

Reply:
Ok, let's leave `Option<dyn Validator>, but still I cannot see what benefit `Fetcher::with_validator` brings rather than adding that function into fetcher constructor. It does not enforce anything new in comparison with passing validator into constructor

Reply:
As I probably said in PMs, the better way to do it is to make `write_executable` a struct, which will store the validator, accept it in contsructor, and pass this struct into the fetchers. But I thought that it's a little over engineering at current point at time and decided to make an easy solution before making something more complex :)